[WIP]Make a common template for /api pages

MSurfer20 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
This PR removes the need for multiple `.md` files and makes a common template to generate /api pages from the OpenAPI data. This is currently a WIP.
**Work plan:**
- [x] Add title of API pages as `summary` parameter in YAML.
- [x] Write MarkdownProcessor to get title from `summary` parameter.
- [x] Refactor titles of existing /api files to use markdowngenerator.
- [x] Fix HTML titles for the /api pages.
- [x] Add the template `.md` file.
- [x] Refactor code to use the template `.md` file.  
- [x] Remove files following the template(CHECKED VIA SCRIPT)
- [x] Proper rendering of JS examples.
- [x] Proper rendering of 404 responses.
- [x] Tackle specific `curl` options
- [x] Tackle specific configurations(like admin).
- [x] Migrate stray descriptions.

**Testing plan:** <!-- How have you tested? -->


**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

timabbott commented on templates/zerver/api/api-doc-template.md:
```diff
@@ -0,0 +1,32 @@
+{start_tabs}
+{tab|python}
+
+{generate_code_example(python)|API_ENDPOINT_NAME|example(admin_config=True)}
```
Why `admin_config=True` here?

MSurfer20 replied:
That was a slight overlook on my part. Fixed.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -10433,6 +10500,7 @@ paths:
   /zulip-outgoing-webhook:
     post:
       operationId: zulip_outgoing_webhooks
+      summary: NO MD FILE.
```
I'd just call this "Outgoing webhooks".

MSurfer20 replied:
Changed to "Outgoing webhooks" :)

timabbott commented on zerver/openapi/markdown_extension.py:
```diff
@@ -485,5 +494,43 @@ def render_description(self, function: str) -> List[str]:
+        title: List[str] = []
+        path, method = function.rsplit(":", 1)
+        title_dict = "# " + get_openapi_summary(path, method)
+        title_dict = title_dict.replace("{{api_url}}", self.api_url)
```
why the variable name `title_Dict`?

MSurfer20 replied:
It was indeed a misleading name. Changed it now.

timabbott commented on templates/zerver/help/include/rest-endpoints.md:
```diff
@@ -10,7 +10,7 @@
 * [Render a message](/api/render-message)
 * [Get a message's raw Markdown](/api/get-raw-message)
-* [Check messages match narrow](/api/check-narrow-matches)
+* [Check messages match narrow](/api/check_messages_match_narrow)
```
We probably want to keep this as `-`s for consistency.

MSurfer20 replied:
Yes. I was going to replace all `_` by `-` in this file, but forgot before pushing.

timabbott commented on zerver/openapi/openapi.py:
```diff
@@ -218,6 +218,19 @@ def get_openapi_summary(endpoint: str, method: str) -> str:
+def get_endpoint_from_operationid(operationid: str) -> str:
+    print(operationid)
+    for each_endpoint in openapi_spec.openapi()["paths"]:
+        for each_method in openapi_spec.openapi()["paths"][each_endpoint]:
```
I'd call these local variables `endpoint` and `method`; no need for the `each_` prefix.

timabbott commented on zerver/openapi/openapi.py:
```diff
@@ -218,6 +218,19 @@ def get_openapi_summary(endpoint: str, method: str) -> str:
+            if (
+                "operationId" in openapi_spec.openapi()["paths"][each_endpoint][each_method]
+                and openapi_spec.openapi()["paths"][each_endpoint][each_method]["operationId"]
+                == operationid
```
It might be cleaner to do `operationId = openapi_spec.openapi()["paths"][endpoint][method].get("operationId")`

and then just compare that value.

timabbott commented on zerver/openapi/openapi.py:
```diff
@@ -218,6 +218,19 @@ def get_openapi_summary(endpoint: str, method: str) -> str:
+                == operationid
+            ):
+                return each_endpoint + ":" + each_method
+    return ""
```
I'd recommend returning a tuple, rather than a `:`-delimited string that the caller needs to parse.

Should we throw an AssertionError if we didn't find it?

MSurfer20 replied:
Changed it to tuple. Also, AssertionError makes sense in such case -- added that

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -4983,7 +4983,7 @@ paths:
     get:
       operationId: get_file_temporary_url
-      summary: NO MD FILE.
+      summary: Get file's temporary URL
```
Possibly just `Get public temporary URL` in an "Files" section makes the most sense.

MSurfer20 commented:
Thanks for the review @timabbott ! I have addressed the comments, and am fixing the currently failing tests. The PR should be ready for another review by tomorrow.

timabbott commented:
This is great @MSurfer20!  I reviewed the first half of the PR and merged it with the following changes.  (Mainly a few different editorial decisions for what to call endpoints with non-standard names).

I'll post a batch of comments next on the remaining changes.

```
diff --git a/templates/zerver/api/changelog.md b/templates/zerver/api/changelog.md
index c461276050..0f591a065a 100644
--- a/templates/zerver/api/changelog.md
+++ b/templates/zerver/api/changelog.md
@@ -162,14 +162,14 @@ field with an integer field `invite_to_realm_policy`.
 
 **Feature level 49**
 
-* Added new [`POST /realm/playground`](/api/add-realm-playground) and
-  [`DELETE /realm/playground/{playground_id}`](/api/remove-realm-playground)
-  endpoints for realm playgrounds.
+* Added new [`POST /realm/playground`](/api/add-code-playground) and
+  [`DELETE /realm/playground/{playground_id}`](/api/remove-code-playground)
+  endpoints for code playgrounds.
 * [`GET /events`](/api/get-events): A new `realm_playgrounds` events
-  is sent when changes are made to a set of configured playgrounds for
+  is sent when changes are made to a set of configured code playgrounds for
   an organization.
 * [`POST /register`](/api/register-queue): Added a new `realm_playgrounds`
-  field, which is required to fetch the set of configured playgrounds for
+  field, which is required to fetch the set of configured code playgrounds for
   an organization.
 
 **Feature level 48**
diff --git a/templates/zerver/api/check-narrow-matches.md b/templates/zerver/api/check-messages-match-narrow.md
similarity index 100%
rename from templates/zerver/api/check-narrow-matches.md
rename to templates/zerver/api/check-messages-match-narrow.md
diff --git a/templates/zerver/help/include/rest-endpoints.md b/templates/zerver/help/include/rest-endpoints.md
index f630c8f9d7..14f161df2f 100644
--- a/templates/zerver/help/include/rest-endpoints.md
+++ b/templates/zerver/help/include/rest-endpoints.md
@@ -10,7 +10,7 @@
 * [Remove an emoji reaction](/api/remove-reaction)
 * [Render a message](/api/render-message)
 * [Get a message's raw Markdown](/api/get-raw-message)
-* [Check messages match narrow](/api/check-messages-match-narrow)
+* [Check if messages match narrow](/api/check-messages-match-narrow)
 * [Get a message's edit history](/api/get-message-history)
 * [Update personal message flags](/api/update-message-flags)
 * [Mark messages as read in bulk](/api/mark-all-as-read)
@@ -61,8 +61,8 @@
 * [Add a linkifier](/api/add-linkifier)
 * [Update a linkifier](/api/update-linkifier)
 * [Remove a linkifier](/api/remove-linkifier)
-* [Add a playground](/api/add-realm-playground)
-* [Remove a playground](/api/remove-realm-playground)
+* [Add a code playground](/api/add-code-playground)
+* [Remove a code playground](/api/remove-code-playground)
 * [Get all custom emoji](/api/get-custom-emoji)
 * [Upload custom emoji](/api/upload-custom-emoji)
 * [Get all custom profile fields](/api/get-custom-profile-fields)
diff --git a/tools/linter_lib/custom_check.py b/tools/linter_lib/custom_check.py
index 89b46cc1cb..3b48a965c3 100644
--- a/tools/linter_lib/custom_check.py
+++ b/tools/linter_lib/custom_check.py
@@ -884,10 +884,7 @@ help_markdown_rules = RuleList(
         {
             "pattern": r"\b[rR]ealm[s]?\b",
             "include_only": {"templates/zerver/help/"},
-            "exclude": {
-                "templates/zerver/help/change-organization-url.md",
-                "templates/zerver/help/include/rest-endpoints.md",
-            },
+            "exclude": {"templates/zerver/help/change-organization-url.md"},
             "good_lines": ["Organization", "deactivate_realm", "realm_filter"],
             "bad_lines": ["Users are in a realm", "Realm is the best model"],
             "description": "Realms are referred to as Organizations in user-facing docs.",
diff --git a/zerver/openapi/openapi.py b/zerver/openapi/openapi.py
index 951d5651d6..f2052bee02 100644
--- a/zerver/openapi/openapi.py
+++ b/zerver/openapi/openapi.py
@@ -224,7 +224,7 @@ def get_endpoint_from_operationid(operationid: str) -> Tuple[str, str]:
             operationId = openapi_spec.openapi()["paths"][endpoint][method].get("operationId")
             if operationId == operationid:
                 return (endpoint, method)
-    raise AssertionError("No such page exists in OpenApi data.")
+    raise AssertionError("No such page exists in OpenAPI data.")
 
 
 def get_openapi_paths() -> Set[str]:
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index dae8854340..f14a638bb9 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -7040,8 +7040,8 @@ paths:
                 $ref: "#/components/schemas/JsonSuccess"
   /realm/playgrounds:
     post:
-      operationId: add_realm_playground
-      summary: Add realm playground
+      operationId: add_code_playground
+      summary: Remove a code playground
       tags: ["server_and_organizations"]
       description: |
         Configure [code playgrounds](/help/code-blocks#code-playgrounds) for the organization.
@@ -7097,8 +7097,8 @@ paths:
                     example: {"id": 1, "result": "success", "msg": ""}
   /realm/playgrounds/{playground_id}:
     delete:
-      operationId: remove_realm_playground
-      summary: Remove realm playground
+      operationId: remove_code_playground
+      summary: Remove a code playground
       tags: ["server_and_organizations"]
       description: |
         Remove a [code playground](/help/code-blocks#code-playgrounds) previously
```

timabbott commented on zerver/templatetags/app_filters.py:
```diff
@@ -167,6 +168,12 @@ def render_markdown_path(
 
+    page_url = context.get("OPEN_GRAPH_URL", "")
+    if markdown_string[0] != "#" and "/api" in page_url:
+        endpoint_name = context["OPEN_GRAPH_URL"].split("/api/")[1].replace("-", "_")
```
I think you probably want to be consistent about whether `/api/` is required here.

Also this logic (and others) crash when you visit `/api/invalid`.  So I think that the `zerver/views/documentation.py` logic which decides whether to render the "404 error page" needs to be doing this parsing, not this code, since otherwise it's possible to reach the AssertionError in `get_endpoint_from_operationid`.

Ideally, we do it in one place, but it's not the end of the world to do it twice.  That said, if we are doing it in more than one place, we should have a function rather than copying this in N places.

timabbott commented on zerver/views/documentation.py:
```diff
@@ -75,6 +75,9 @@ def get_path(self, article: str) -> Tuple[str, int]:
             http_status = 404
 
         path = self.path_template % (article,)
+        article_path = os.path.join(settings.DEPLOY_ROOT, "templates") + path
```
Why is path added rather than included in the `os.path.join`?

MSurfer20 replied:
This is later used in the code as well. Quoting from there:
```
# The following is a somewhat hacky approach to extract titles from articles.
        # Hack: `context["article"] has a leading `/`, so we use + to add directories.
        article_path = os.path.join(settings.DEPLOY_ROOT, "templates") + context["article"]
```

timabbott replied:
I see, OK; worth doing a comment then, because generally one should use the preferred combination functions.

(Maybe we want `urljoin` instead?)

MSurfer20 replied:
`urljoin` can only join two at a time, so anyway we would have to use multiple instances of it or combine it with `path.join`. Idts it's as clean a solution as we would expect.

timabbott commented on zerver/tests/test_docs.py:
```diff
@@ -106,17 +107,19 @@ def _test(
-        files = list(filter(_filter_func, files))
+        docs_file = os.path.join(current_dir, "..", "..", "templates/zerver/api/sidebar_index.md")
+        docs_list = open(docs_file, "r").read()
+        endpoint_list = endpoint_list + re.findall(ENDPOINT_REGEXP, docs_list)
```
I think we want to do just do a `client_get` to fetch the actual HTML page, and parse the links on the page.  

(Since we may be trying to auto-generate `rest-endpoints.md` in the future, too)

MSurfer20 replied:
I thought about it, but if we use the client to get the actual raw HTML, then there is no way of distinguishing sidebar links and the main rest-endpoints links. Distinguishing the two kinds of links might make the code messy.

MSurfer20 replied:
I realised that we need to(and we are) test those as well. Changed the code to use the HTML page for parsing links.

timabbott commented:
One other small thought -- you should adjust the commit order so that each commit works.  E.g. move the commit to change discovery tests earlier in the branch, since that one will work both before and after deleting the files.  That makes it easier to safely incrementally merge things.

MSurfer20 commented:
> This is great @MSurfer20! I reviewed the first half of the PR and merged it with the following changes. (Mainly a few different editorial decisions for what to call endpoints with non-standard names).
> 
> I'll post a batch of comments next on the remaining changes.
> 
> ```
> diff --git a/templates/zerver/api/changelog.md b/templates/zerver/api/changelog.md
> index c461276050..0f591a065a 100644
> --- a/templates/zerver/api/changelog.md
> +++ b/templates/zerver/api/changelog.md
> @@ -162,14 +162,14 @@ field with an integer field `invite_to_realm_policy`.
>  
>  **Feature level 49**
>  
> -* Added new [`POST /realm/playground`](/api/add-realm-playground) and
> -  [`DELETE /realm/playground/{playground_id}`](/api/remove-realm-playground)
> -  endpoints for realm playgrounds.
> +* Added new [`POST /realm/playground`](/api/add-code-playground) and
> +  [`DELETE /realm/playground/{playground_id}`](/api/remove-code-playground)
> +  endpoints for code playgrounds.
>  * [`GET /events`](/api/get-events): A new `realm_playgrounds` events
> -  is sent when changes are made to a set of configured playgrounds for
> +  is sent when changes are made to a set of configured code playgrounds for
>    an organization.
>  * [`POST /register`](/api/register-queue): Added a new `realm_playgrounds`
> -  field, which is required to fetch the set of configured playgrounds for
> +  field, which is required to fetch the set of configured code playgrounds for
>    an organization.
>  
>  **Feature level 48**
> diff --git a/templates/zerver/api/check-narrow-matches.md b/templates/zerver/api/check-messages-match-narrow.md
> similarity index 100%
> rename from templates/zerver/api/check-narrow-matches.md
> rename to templates/zerver/api/check-messages-match-narrow.md
> diff --git a/templates/zerver/help/include/rest-endpoints.md b/templates/zerver/help/include/rest-endpoints.md
> index f630c8f9d7..14f161df2f 100644
> --- a/templates/zerver/help/include/rest-endpoints.md
> +++ b/templates/zerver/help/include/rest-endpoints.md
> @@ -10,7 +10,7 @@
>  * [Remove an emoji reaction](/api/remove-reaction)
>  * [Render a message](/api/render-message)
>  * [Get a message's raw Markdown](/api/get-raw-message)
> -* [Check messages match narrow](/api/check-messages-match-narrow)
> +* [Check if messages match narrow](/api/check-messages-match-narrow)
>  * [Get a message's edit history](/api/get-message-history)
>  * [Update personal message flags](/api/update-message-flags)
>  * [Mark messages as read in bulk](/api/mark-all-as-read)
> @@ -61,8 +61,8 @@
>  * [Add a linkifier](/api/add-linkifier)
>  * [Update a linkifier](/api/update-linkifier)
>  * [Remove a linkifier](/api/remove-linkifier)
> -* [Add a playground](/api/add-realm-playground)
> -* [Remove a playground](/api/remove-realm-playground)
> +* [Add a code playground](/api/add-code-playground)
> +* [Remove a code playground](/api/remove-code-playground)
>  * [Get all custom emoji](/api/get-custom-emoji)
>  * [Upload custom emoji](/api/upload-custom-emoji)
>  * [Get all custom profile fields](/api/get-custom-profile-fields)
> diff --git a/tools/linter_lib/custom_check.py b/tools/linter_lib/custom_check.py
> index 89b46cc1cb..3b48a965c3 100644
> --- a/tools/linter_lib/custom_check.py
> +++ b/tools/linter_lib/custom_check.py
> @@ -884,10 +884,7 @@ help_markdown_rules = RuleList(
>          {
>              "pattern": r"\b[rR]ealm[s]?\b",
>              "include_only": {"templates/zerver/help/"},
> -            "exclude": {
> -                "templates/zerver/help/change-organization-url.md",
> -                "templates/zerver/help/include/rest-endpoints.md",
> -            },
> +            "exclude": {"templates/zerver/help/change-organization-url.md"},
>              "good_lines": ["Organization", "deactivate_realm", "realm_filter"],
>              "bad_lines": ["Users are in a realm", "Realm is the best model"],
>              "description": "Realms are referred to as Organizations in user-facing docs.",
> diff --git a/zerver/openapi/openapi.py b/zerver/openapi/openapi.py
> index 951d5651d6..f2052bee02 100644
> --- a/zerver/openapi/openapi.py
> +++ b/zerver/openapi/openapi.py
> @@ -224,7 +224,7 @@ def get_endpoint_from_operationid(operationid: str) -> Tuple[str, str]:
>              operationId = openapi_spec.openapi()["paths"][endpoint][method].get("operationId")
>              if operationId == operationid:
>                  return (endpoint, method)
> -    raise AssertionError("No such page exists in OpenApi data.")
> +    raise AssertionError("No such page exists in OpenAPI data.")
>  
>  
>  def get_openapi_paths() -> Set[str]:
> diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
> index dae8854340..f14a638bb9 100644
> --- a/zerver/openapi/zulip.yaml
> +++ b/zerver/openapi/zulip.yaml
> @@ -7040,8 +7040,8 @@ paths:
>                  $ref: "#/components/schemas/JsonSuccess"
>    /realm/playgrounds:
>      post:
> -      operationId: add_realm_playground
> -      summary: Add realm playground
> +      operationId: add_code_playground
> +      summary: Remove a code playground
>        tags: ["server_and_organizations"]
>        description: |
>          Configure [code playgrounds](/help/code-blocks#code-playgrounds) for the organization.
> @@ -7097,8 +7097,8 @@ paths:
>                      example: {"id": 1, "result": "success", "msg": ""}
>    /realm/playgrounds/{playground_id}:
>      delete:
> -      operationId: remove_realm_playground
> -      summary: Remove realm playground
> +      operationId: remove_code_playground
> +      summary: Remove a code playground
>        tags: ["server_and_organizations"]
>        description: |
>          Remove a [code playground](/help/code-blocks#code-playgrounds) previously
> ```

Thanks for the merge @timabbott ! In this change, I notice that there was a typo of 'Add' being written as 'Remove' for the `add_code_playground` summary. Opened https://github.com/zulip/zulip/pull/18625/ to fix it :)

MSurfer20 commented:
@timabbott I have replied to most of the comments above. Further, the deleted api pages currently are the ones where there is no diff between the original page and page after deleting. However, this removed some pages that earlier i thought would be exact matches--so I'll have to look more into this.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -126,7 +126,7 @@ paths:
       responses:
         "200":
-          description: Success.
+          description: A typical successful JSON response may look like
```
Can we pull this description from a common entry rather than having N copies of it?

MSurfer20 replied:
That is what I was gonna do in the next commit :)(along with fixing the cases where the script didn't find any descriptions).

timabbott commented on zerver/views/documentation.py:
```diff
@@ -75,28 +75,52 @@ def get_path(self, article: str) -> Tuple[str, int]:
+                http_status_ignored,
+                endpoint_name_ignored,
+                endpoint_method_ignored,
+            ) = self.get_path("sidebar_index")
```
This calling convention is really messy.  Rather than returning N variables, can you define a `@dataclass` for the return value of `self.get_path` in a prep commit (with no functional changes) with the `sidebar_index` and `http_status` fields, and then later commits can extend that object?  

I'm also not sure we need to modify `get_path` in the /help/ code path at all; could we instead just have the /api/ code be in a function only present on that subclass?

MSurfer20 replied:
Changed  to use the dataclass.

Splitting the two functions would've duplicated a lot of code IMO, since the endpoint_name and endpoin_method are coming as a by-product of checking the http_status for the endpoint(to differentiate between an invalid endpoint where we return 404, or a valid endpoint whose file got deleted, in case we return 200).

timabbott commented on zerver/tests/test_docs.py:
```diff
@@ -105,20 +106,22 @@ def _test(
+            follow=True,
+            HTTP_X_REQUESTED_WITH="XMLHttpRequest",
+        )
+        self.assertEqual(result.status_code, 404)
```
Can you pull this commit as well as a targeted bug fix for /api/invalid being broken in master into a smaller prep PR that we can merge now?  We're getting 500s in chat.zulip.org from crawlers right now, and I'd like to be able to merge a fix for that (with this test code) tomorrow even if the rest needs more round trips.

timabbott commented on zerver/openapi/test_curl_examples.py:
```diff
@@ -35,45 +37,67 @@ def test_generated_curl_examples_for_success(client: Client) -> None:
+    endpoint_list = sorted(re.findall(ENDPOINT_REGEXP, api_docs_list))
+    for endpoint in endpoint_list:
+        article_name = endpoint + ".md"
+        file_path = os.path.join(current_dir, "..", "..", "templates/zerver/api/", article_name)
```
Why the change to use the `../../` stuff?  We previously were having paths starting from the CWD, which should be correct and save.

MSurfer20 replied:
@timabbott opened https://github.com/zulip/zulip/pull/18696 for this commit instead.

timabbott commented on zerver/openapi/markdown_extension.py:
```diff
@@ -154,7 +154,8 @@ def render_javascript_code_example(
         m = re.search(pattern, f.read(), re.M | re.S)
-    assert m is not None
+    if m is None:
+        return []
```
Can you add a comment explaining the purpose of this case?

MSurfer20 replied:
The usefulness of this comes in the form that we can add the lines to generate JS examples whether that particular endpoint actually has JS examples or not. Thus, it would expand coverage of the template. Maybe I should split this into a different commit to make the purpose more clear?

timabbott replied:
Yeah, seems wise.

MSurfer20 replied:
@timabbott Hmmm I'll split these further. At this point, the first 4 commits(till `apidocs: Remove the pages that follow the standard template`) are pretty much ready for merge. The remaining would still need some work as of now.

timabbott commented on zerver/openapi/markdown_extension.py:
```diff
@@ -164,7 +165,11 @@ def render_javascript_code_example(
+    code_example = [
+        "{tab|js}\n",
+        "More examples and documentation can be found [here](https://github.com/zulip/zulip-js).",
+    ]
```
I think this block may need to be in the next commit, to avoid double-including this content.

MSurfer20 replied:
@timabbott Indeed this breaks without the second commit. I had split it for easier review in this case.

timabbott commented on zerver/views/documentation.py:
```diff
@@ -86,26 +86,37 @@ def get_path(self, article: str) -> Tuple[str, int]:
         context: Dict[str, Any] = super().get_context_data()
-        (context["article"], http_status_ignored) = self.get_path(article)
+
+        Articlepath: ArticlePath = self.get_path(article)
```
This is very weird code; you should write `article_path = self.get_path(article)`, and mypy should infer the type.

MSurfer20 replied:
@timabbott Changed the code to the other format.

MSurfer20 commented:
Attaching the diff of all the files. Of these, only `update_display_settings` and `update_notification_settings` need to be looked at, since their return values weren't displayed earlier, but now they appear due to the template. Others are just improvements over what the templates displayed.
```
old-check-messages-match-narrow new-check-messages-match-narrow
262c262
< <h2 id="arguments">Arguments</h2>
---
> <h2 id="parameters">Parameters</h2>
old-check-messages-match-narrow new-check-messages-match-narrow

old-fetch-api-key new-fetch-api-key
280c280
< <h3 id="example-response">Example response</h3>
---
> <h4 id="example-response">Example response</h4>
old-fetch-api-key new-fetch-api-key

old-get-own-user new-get-own-user
263c263
< <p>This endpoint takes no parameters.</p>
---
> <p>This endpoint does not accept any parameters.</p>
old-get-own-user new-get-own-user

old-get-subscription-status new-get-subscription-status
269a270,273
> <h4 id="return-values">Return values</h4>
> <ul>
> <li><code>is_subscribed</code>: <span class="api-field-type">boolean</span> Whether the user is subscribed to the stream.</li>
> </ul>
old-get-subscription-status new-get-subscription-status

old-unsubscribe new-unsubscribe
310a311
> <h2 id="response">Response</h2>
old-unsubscribe new-unsubscribe

old-update-display-settings new-update-display-settings
409a410,458
> <ul>
> <li>
> <p><code>twenty_four_hour_time</code>: <span class="api-field-type">boolean</span> The setting for <code>twenty_four_hour_time</code>, if it was changed in this request.</p>
> </li>
> <li>
> <p><code>dense_mode</code>: <span class="api-field-type">boolean</span> The setting for <code>dense_mode</code>, if it was changed in this request.
>     This setting is however reserved for future, and can not be
>     modified in production environment.</p>
> </li>
> <li>
> <p><code>starred_message_counts</code>: <span class="api-field-type">boolean</span> The setting for <code>starred_message_counts</code>, if it was changed
>     in this request.</p>
> </li>
> <li>
> <p><code>fluid_layout_width</code>: <span class="api-field-type">boolean</span> The setting for <code>fluid_layout_width</code>, if it was changed
>     in this request.</p>
> </li>
> <li>
> <p><code>high_contrast_mode</code>: <span class="api-field-type">boolean</span> The setting for <code>high_contrast_mode</code>, if it was changed in
>     this request.
>     This setting is however reserved for future, and can not be
>     modified in production environment.</p>
> </li>
> <li>
> <p><code>color_scheme</code>: <span class="api-field-type">integer</span> The numerical key corresponding to new <code>color_scheme</code> if it was changed in this request.</p>
> </li>
> <li>
> <p><code>translate_emoticons</code>: <span class="api-field-type">boolean</span> The setting for <code>translate_emoticons</code>, if it was changed in this request.</p>
> </li>
> <li>
> <p><code>default_language</code>: <span class="api-field-type">string</span> The language code corresponding to new <code>default_language</code> if it was changed in this request.</p>
> </li>
> <li>
> <p><code>default_view</code>: <span class="api-field-type">string</span> The new setting for <code>default_view</code>, if it was changed in this request.</p>
> </li>
> <li>
> <p><code>left_side_userlist</code>: <span class="api-field-type">boolean</span> The setting for <code>left_side_userlist</code>, if it was changed in this request.</p>
> </li>
> <li>
> <p><code>emojiset</code>: <span class="api-field-type">string</span> The string identifier corresponding to new <code>emojiset</code> if it was changed in this request.</p>
> </li>
> <li>
> <p><code>demote_inactive_streams</code>: <span class="api-field-type">integer</span> The numerical key corresponding to new <code>demote_inactive_streams</code> setting if it was changed in this request.</p>
> </li>
> <li>
> <p><code>timezone</code>: <span class="api-field-type">string</span> The setting for <code>timezone</code>, if it was changed in this request.</p>
> </li>
> </ul>
> <h4 id="return-values_1">Return values</h4>
old-update-display-settings new-update-display-settings

old-update-notification-settings new-update-notification-settings
436a437,482
> <ul>
> <li>
> <p><code>enable_desktop_notifications</code>: <span class="api-field-type">boolean</span> The setting for <code>enable_desktop_notifications</code>, if it was changed in
>     this request.</p>
> </li>
> <li>
> <p><code>enable_digest_emails</code>: <span class="api-field-type">boolean</span> The setting for <code>enable_digest_emails</code>, if it was changed in this
>     request.</p>
> </li>
> <li>
> <p><code>enable_marketing_emails</code>: <span class="api-field-type">boolean</span> The setting for <code>enable_marketing_emails</code>, if it was changed in this
>     request.</p>
> </li>
> <li>
> <p><code>enable_offline_email_notifications</code>: <span class="api-field-type">boolean</span> The setting for <code>enable_offline_email_notifications</code>, if it was changed
>     in this request.</p>
> </li>
> <li>
> <p><code>enable_offline_push_notifications</code>: <span class="api-field-type">boolean</span> The setting for <code>enable_offline_push_notifications</code>, if it was changed
>     in this request.</p>
> </li>
> <li>
> <p><code>enable_online_push_notifications</code>: <span class="api-field-type">boolean</span> The setting for <code>enable_online_push_notifications</code>, if it was changed in
>     this request.</p>
> </li>
> <li>
> <p><code>enable_sounds</code>: <span class="api-field-type">boolean</span> The setting for <code>enable_sounds</code>, if it was changed in this request.</p>
> </li>
> <li>
> <p><code>enable_stream_email_notifications</code>: <span class="api-field-type">boolean</span> The setting for <code>enable_stream_email_notifications</code>, if it was changed
>     in this request.</p>
> </li>
> <li>
> <p><code>enable_stream_push_notifications</code>: <span class="api-field-type">boolean</span> The setting for <code>enable_stream_push_notifications</code>, if it was changed in
>     this request.</p>
> </li>
> <li>
> <p><code>enable_stream_audible_notifications</code>: <span class="api-field-type">boolean</span> The setting for <code>enable_stream_audible_notifications</code>, if it was changed
>     in this request.</p>
> </li>
> <li>
> <p><code>message_content_in_email_notifications</code>: <span class="api-field-type">boolean</span> The setting for <code>message_content_in_email_notifications</code>, if it was
>     changed in this request.</p>
> </li>
> </ul>
> <h4 id="return-values_1">Return values</h4>
old-update-notification-settings new-update-notification-settings

old-upload-file new-upload-file
261a262
> <p>This endpoint does not accept any parameters.</p>
old-upload-file new-upload-file
```

MSurfer20 commented:
Also, talking about the files which i couldn't remove as of now(I would appreciate feedback about how I should proceed with them):
* construct-narrow: This one is all text, and doesn't remotely follow the template.
* create-stream:  This one is all text, and doesn't remotely follow the template.
* get-events: This one contains `import sys` in its python example, due to which it is currently hardcoded in the template.
* mark-all-as-read: This one contains 3 endpoints documentation. We can split it to make 3 different doc pages, but these do seem to fit better on a single page.
* real-time-events: This one contains `import sys` in its python example, due to which it is currently hardcoded in the template.
* register-queue: ?(I need to cross-check this one).
* send-message: Its docs also contain zulip-send and bash examples, and so it can't be fit into the template.

MSurfer20 commented:
@timabbott Managed to delete 58 files :)

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -10044,6 +10044,13 @@ paths:
             type: boolean
           example: true
+      x-response-description: |
+        #### Return values
```
I think we don't want this heading along, since it's a duplicate.

MSurfer20 replied:
This already got merged with the changed description :)

timabbott commented on zerver/lib/markdown/api_return_values_table_generator.py:
```diff
@@ -45,7 +45,9 @@ def run(self, lines: List[str]) -> List[str]:
+                # delete-queue is not supposed to have return
+                # values.
+                elif doc_name != "/events:delete":
                     return_values = copy.deepcopy(return_values)
```
I don't understand this logic; shouldn't it just be `if doc_name != "/events:get" / else` , so that every endpoint but `events:get` has the default logic?

(I think also a cleaner way to write it would be to have `if doc_name == "/events:get": / else` block.)

timabbott commented on zerver/openapi/markdown_extension.py:
```diff
@@ -129,6 +129,8 @@ def extract_code_example(
     function: str, admin_config: bool = False, **kwargs: Any
 ) -> List[str]:
+    if function not in zerver.openapi.python_examples.TEST_FUNCTIONS:
+        return []
```
Do we consider it a bug to have any endpoint without a Python code example?  If so, we could consider `raise AssertionError()` here, rather than `return []`, so that it's not a silent failure.

MSurfer20 replied:
Currently, the two fetch-api-keys endpoints don't have python examples -- hence the empty string.

timabbott replied:
OK, cool.  I think we probably want to add them and then make this obligatory in a follow-up PR.

timabbott commented:
I merged 4 early commits as the series ending with e892a02bb8e1f06fa728c70f2fe37d7ec069376c.

I agree that the output looks great; essentially all the changes are bug fixes :).  Posted a comment on the Python commit, but I don't think it's a blocker for merging (since we can always fix that forward).  

I think I agree with your thoughts on which of these can be readily deleted; we should fix the `import sys` thing and let the rest be.

timabbott commented on zerver/views/documentation.py:
```diff
@@ -86,26 +86,37 @@ def get_path(self, article: str) -> Tuple[str, int]:
         context: Dict[str, Any] = super().get_context_data()
-        (context["article"], http_status_ignored) = self.get_path(article)
+
+        Article_path = self.get_path(article)
```
This should be `article_path = `; we never use capital letters in local variable names unless the name is an ALL_CAPS constant/settings.

MSurfer20 replied:
Changed the naming.

timabbott commented on zerver/views/documentation.py:
```diff
@@ -19,6 +20,14 @@
+    article: str
+    article_http_status: int
+    endpoint_name: str
+    endpoint_method: str
```
I kinda feel like this name isn't quite right; maybe we want:
```
class DocumentationArticle:
    article_path: str
    article_http_status: int
    endpoint_path: Optional[str]  # should be `None` rather than `""` if not a REST API page?
    endpoint_method: Optional[str]  # should be `None` rather than `""` if not a REST API page?
```

(I think `endpoint_path`, since `endpoint_name` would be the `operationId`, not the thing with `/`s in it).

MSurfer20 replied:
Changed the naming.

MSurfer20 commented:
@timabbott Apologies for the unnoticed bug.... Didn't realise it since all the tests had passed in the end.

timabbott commented on zerver/tests/test_docs.py:
```diff
@@ -142,7 +142,7 @@ def test_doc_endpoints(self) -> None:
         self._test("/api/delete-queue", "Delete a previously registered queue")
         self._test("/api/update-message", "propagate_mode")
-        self._test("/api/get-own-user", "takes no parameters")
+        self._test("/api/get-own-user", "does not accept any parameters.")
```
We don't want to have a commit where master fails tests, which the previous setup arranges.  We can address this by just moving this before the previous commit, and manually changing that file to use the standard string.  I'll do that while merging.

timabbott commented:
Merged after the small update above and also squashing the "add template" and "use template" commits.  Huge thanks for doing this giant migration @MSurfer20!

A few notes on near-term follow-ups:
* We should probably switch to consistently using `endpoint_path` as the local variable name for e.g. `/messages/{message_id}` URL paths; `endpoint_name` is confusing for something with `/`s in it.  It's still a little mixed right now.
* We should do an update of the "writing API documentation" docs, since there's a lot of things one no longer needs to do.  We can close https://github.com/zulip/zulip/issues/12571 after doing so.
* It's worth noting that `/api/get_messsages` now works just as well as `/api/get-messages`.  That's probably a good thing; I remember someone wanting that feature.  I'll post about it in `#documentation`.
